File: db/rag-indexer.ts
import fs from "node:fs";
import path from "node:path";
import postgres from "postgres";
import { chunkText, computeHash, generateEmbeddings } from "./rag-service";

const databaseUrl = process.env.DATABASE_URL;

// Helper to URL-encode password in database URL to prevent connection string parsing errors
function safePostgresClient() {
  if (!databaseUrl) return null;
  
  if (!databaseUrl.startsWith("postgresql://") && !databaseUrl.startsWith("postgres://")) {
    return postgres(databaseUrl);
  }
  
  const prefix = databaseUrl.startsWith("postgresql://") ? "postgresql://" : "postgres://";
  const rest = databaseUrl.slice(prefix.length);
  const lastAtIndex = rest.lastIndexOf("@");
  if (lastAtIndex === -1) return postgres(databaseUrl);
  
  const creds = rest.slice(0, lastAtIndex);
  const hostPart = rest.slice(lastAtIndex + 1);
  
  const colonIndex = creds.indexOf(":");
  if (colonIndex === -1) return postgres(databaseUrl);
  
  const user = creds.slice(0, colonIndex);
  const pass = creds.slice(colonIndex + 1);
  
  const safeUrl = `${prefix}${encodeURIComponent(user)}:${encodeURIComponent(pass)}@${hostPart}`;
  return postgres(safeUrl);
}

const sql = safePostgresClient();

// Files and folders to ignore during indexing
const EXCLUDE_DIRS = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  ".bun",
  "todo-app",
  "todo-list-app",
]);

const INCLUDE_EXTENSIONS = new Set([
  ".ts",
  ".tsx",
  ".js",
  ".jsx",
  ".json",
  ".md",
  ".css",
  ".html",
  ".yaml",
  ".yml",
]);

interface PendingChunk {
  filePath: string;
  content: string;
  index: number;
  hash: string;
}

export async function indexCodebase(codebasePath: string) {
  if (!sql) {
    console.error("❌ Database is not connected.");
    return;
  }

  console.log(`\n🔍 Scanning codebase for files in: ${codebasePath}`);
  const filesToIndex: string[] = [];

  function walk(dir: string) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const ent of entries) {
      const fullPath = path.join(dir, ent.name);
      const relPath = path.relative(codebasePath, fullPath);

      if (ent.isDirectory()) {
        if (EXCLUDE_DIRS.has(ent.name)) continue;
        walk(fullPath);
      } else {
        const ext = path.extname(ent.name).toLowerCase();
        if (INCLUDE_EXTENSIONS.has(ext)) {
          filesToIndex.push(relPath);
        }
      }
    }
  }

  walk(codebasePath);
  console.log(`📂 Found ${filesToIndex.length} code files to check.`);

  let skippedCount = 0;
  let updatedCount = 0;
  const pendingChunks: PendingChunk[] = [];

  for (const relPath of filesToIndex) {
    const absPath = path.resolve(codebasePath, relPath);
    const content = fs.readFileSync(absPath, "utf8");
    const fileHash = computeHash(content);

    // 1. Check if the file hash is already indexed and matches
    const existing = await sql`
      SELECT id FROM code_chunks 
      WHERE file_path = ${relPath} AND file_hash = ${fileHash} 
      LIMIT 1
    `;

    if (existing.length > 0) {
      skippedCount++;
      continue;
    }

    updatedCount++;
    // File changed, delete old chunks from the database
    await sql`
      DELETE FROM code_chunks WHERE file_path = ${relPath}
    `;

    const chunks = chunkText(content);
    chunks.forEach((chunkContent, idx) => {
      pendingChunks.push({
        filePath: relPath,
        content: chunkContent,
        index: idx,
        hash: fileHash,
      });
    });
  }

  console.log(`⏭️  Unchanged files skipped: ${skippedCount}`);
  console.log(`📝 Files to index/update: ${updatedCount} (${pendingChunks.length} total chunks)`);

  if (pendingChunks.length === 0) {
    console.log("✨ Codebase is already up to date. No indexing required.");
    await sql.end();
    return;
  }

  // 2. Batch embedding calls and database insertions
  const BATCH_SIZE = 40;
  let processedChunks = 0;

  console.log(`🚀 Starting batch indexing (Batch size: ${BATCH_SIZE})...`);

  for (let i = 0; i < pendingChunks.length; i += BATCH_SIZE) {
    const batch = pendingChunks.slice(i, i + BATCH_SIZE);
    const batchTexts = batch.map((c) => c.content);

    try {
      console.log(`  ⚡ Generating embeddings for chunks ${i + 1} to ${Math.min(i + BATCH_SIZE, pendingChunks.length)}...`);
      const embeddings = await generateEmbeddings(batchTexts);

      console.log(`  💾 Inserting chunks into database...`);
      // Insert in a transaction to be safe and fast
      await sql.begin(async (tx) => {
        for (let j = 0; j < batch.length; j++) {
          const chunk = batch[j]!;
          const embedding = embeddings[j]!;
          const vectorStr = `[${embedding.join(",")}]`;

          await tx`
            INSERT INTO code_chunks (file_path, content, chunk_index, file_hash, embedding)
            VALUES (${chunk.filePath}, ${chunk.content}, ${chunk.index}, ${chunk.hash}, ${vectorStr}::vector)
          `;
        }
      });

      processedChunks += batch.length;
    } catch (err) {
      console.error(`❌ Error indexing batch starting at index ${i}:`, err);
    }
  }

  console.log(`\n🎉 Indexing finished successfully!`);
  console.log(`   Processed ${processedChunks}/${pendingChunks.length} chunks across ${updatedCount} files.`);
  console.log(`   Skipped ${skippedCount} unchanged files.`);

  await sql.end();
}

File: python/db/rag_indexer.py
"""
db/rag_indexer.py — Incremental codebase indexer using LangChain + SHA-256 hashing.

Usage:
    python python/main.py index
"""
from __future__ import annotations

import hashlib
import os
from pathlib import Path

from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_core.documents import Document

from config import CODEBASE_PATH
from db.rag_service import get_vector_store
from db.supabase_client import get_supabase

# Directories to skip during crawl
EXCLUDE_DIRS = {
    "node_modules", ".git", "dist", "build", ".bun",
    "todo-app", "todo-list-app", "__pycache__", ".mypy_cache",
    ".venv", "venv", "env",
}

# Source file extensions to index
INCLUDE_EXTENSIONS = {
    ".ts", ".tsx", ".js", ".jsx", ".json", ".md",
    ".css", ".html", ".yaml", ".yml", ".py", ".toml",
}

CHUNK_SIZE = 1000
CHUNK_OVERLAP = 200
BATCH_SIZE = 40


def _compute_hash(content: str) -> str:
    return hashlib.sha256(content.encode("utf-8")).digest().hex()


def _walk_codebase(root: str) -> list[Path]:
    """Recursively collect indexable source files."""
    files: list[Path] = []
    for dirpath, dirnames, filenames in os.walk(root):
        # Prune excluded directories in-place so os.walk doesn't descend into them
        dirnames[:] = [d for d in dirnames if d not in EXCLUDE_DIRS]
        for fname in filenames:
            if Path(fname).suffix.lower() in INCLUDE_EXTENSIONS:
                files.append(Path(dirpath) / fname)
    return files


def index_codebase(codebase_path: str = CODEBASE_PATH) -> None:
    vs = get_vector_store()
    sb = get_supabase()

    if not vs or not sb:
        print("❌ Vector store or Supabase client is not initialised.")
        return

    splitter = RecursiveCharacterTextSplitter(
        chunk_size=CHUNK_SIZE,
        chunk_overlap=CHUNK_OVERLAP,
    )

    print(f"\n🔍 Scanning codebase: {codebase_path}")
    all_files = _walk_codebase(codebase_path)
    print(f"📂 Found {len(all_files)} source files to check.\n")

    pending_docs: list[Document] = []
    skipped = 0
    updated = 0

    for fp in all_files:
        rel_path = str(fp.relative_to(codebase_path))
        try:
            content = fp.read_text(encoding="utf-8", errors="ignore")
        except Exception:
            continue

        file_hash = _compute_hash(content)

        # Incremental check: skip if file hash already exists in documents table
        try:
            existing = (
                sb.table("documents")
                .select("id")
                .eq("metadata->>file_path", rel_path)
                .eq("metadata->>file_hash", file_hash)
                .limit(1)
                .execute()
            )
            if existing.data:
                skipped += 1
                continue
        except Exception:
            pass  # If check fails, reindex anyway

        # Delete old chunks for this file before reinserting
        try:
            sb.table("documents").delete().eq("metadata->>file_path", rel_path).execute()
        except Exception:
            pass

        updated += 1
        chunks = splitter.split_text(content)
        for idx, chunk in enumerate(chunks):
            pending_docs.append(
                Document(
                    page_content=chunk,
                    metadata={"file_path": rel_path, "chunk_index": idx, "file_hash": file_hash},
                )
            )

    print(f"⏭️  Unchanged files skipped: {skipped}")
    print(f"📝 Files to index/update: {updated} ({len(pending_docs)} total chunks)")

    if not pending_docs:
        print("✨ Codebase is already up to date. No indexing required.")
        return

    print(f"🚀 Starting batch embedding (batch size: {BATCH_SIZE})...")
    for i in range(0, len(pending_docs), BATCH_SIZE):
        batch = pending_docs[i : i + BATCH_SIZE]
        end_idx = min(i + BATCH_SIZE, len(pending_docs))
        print(f"  ⚡ Embedding chunks {i + 1}–{end_idx}...")
        try:
            vs.add_documents(batch)
            print(f"  💾 Inserted {len(batch)} chunks.")
        except Exception as exc:
            print(f"  ❌ Batch failed: {exc}")

    print(f"\n🎉 Indexing done! Processed {len(pending_docs)} chunks across {updated} files.")
    print(f"   Skipped {skipped} unchanged files.")
